### PR TITLE
Relax tolerance in SphericalManifold

### PR DESCRIPTION
**Pull request by masterleinad**
I noticed (as described in #5506) that for a `HyperShell<3>` mesh with radii .5 and 1. we actually only improve rather coarse cells at all and simply end up with noting that the first iteration already succeeded.

The maximum squared (Euclidian) distance between the candidate and the surrounding seems to be a good indicator to break even earlier. 
For the points that need two iterations this threshold is at least .09 (for all refinements) while it is less than .03 for all points in the third refinement that only need one iteration. Therefore .01 seems to be a good compromise which should also make sure that we perform the Newton iteration only if really necessary in other cases as well.

@gassmoeller Can you check that this improves your timings?

### CODE REVIEW

**Comment by kronbichler** [comment]
This looks good to me in principle. The three questions/concerns I have are the following:
- Do I understand you right that "terminate in the first iteration" on level 3 there is almost no point that comes beyond the initial check of the Newton residual, i.e., the residual is less than `1e-10` for the initial tangential approximation? Can we somehow quantify why a distance of `1e-2` translates to this case? My understanding is that we construct a tangential approximation around the points that should be 2nd order accurate. Since we are then blending points at a distance of at most `1e-2` apart, we get another linear term that should make it `(1e-2)^3`. In case we cannot establish quantitative thresholds, it might be better to use a parameter for this setting (or, as @gassmoeller suggested, a `fast` variant vs an `accurate` one).
- As compared to before, the new method does not even enter the first round of computing the trigonometric functions for the minimization, which represents the main saving. Is this correct? Or do we use the first update?
- While the shortcut certainly looks promising, I am a bit concerned about the different levels of accuracy we have inside this class. The problematic case I fixed in #5507 appeared because we had `1e-9` difference in points that should have been along the same line from the center, which appeared due to the tolerance level `1e-10` of the Newton iteration. If we now introduce yet another length scale, say at around `1e-7`, this will stress other downstream uses. As an example, `TransfiniteInterpolationManifold` does finite differencing on the point of a `SphericalManifold`. The usual case, and the only one that is covered by our tests, is to have points farther than `1e-2` apart (we talk about coarse cells in that case), but I don't want to imagine what happens if a user sometimes triggers the fast path and sometimes not.

**Comment by masterleinad** [comment]
> As compared to before, the new method does not even enter the first round of computing the trigonometric functions for the minimization, which represents the main saving. Is this correct? Or do we use the first update?

The whole idea of this is that we don't enter the loop when we expect that the first update is already really close to the original candidate. As compared before, we don't compute the first update which is the main saving.

File: source/grid/manifold_lib.cc [hunk 1]
```diff
@@ -462,6 +465,13 @@ get_new_point (const ArrayView<const Point<spacedim>> &vertices,
           }
       }
 
+    // If all the points are close to the candidate, we expect the candidate to
+    // be good enough. This tolerance was chosen such that the first iteration
+    // for a at least three time refined HyperShell mesh with radii .5 and 1.
+    // doesn't already succeed.
+    if (max_distance < 1.e-2)
+      return center + rho * candidate;
```
Comment (bangerth) [inline source/grid/manifold_lib.cc]: This can't be the correct check -- `max_distance` has been computed in real space (not on the reference cell), i.e., it is sensitive to the size of the domain. Wouldn't this check, for example, always succeed if we are on a sphere of radius 1e-3?

Comment (masterleinad) [reply source/grid/manifold_lib.cc]: Both `candidate` (in line 406) and the `direction`s (in line 437) are normalized. This should make the check independent of the radius as far as I see.

Comment (bangerth) [reply source/grid/manifold_lib.cc]: Oh, I missed this -- I had looked for something that pulls these points back to the reference cell. Thanks for pointing out my mistake. Would you mind adding a parenthetical remark to the comment that mentions that `max_distance` is the distance between points on the *unit* sphere, so an absolute tolerance is ok?

**Comment by gassmoeller** [comment]
Hmm, interestingly this does not change my timings much for a hyper_shell with 3 or 4 refinements (a few percent faster than the current master, but still significantly slower than before). I am travelling and currently have no access to the detailed callgraph, but will try to take a look end of this week. The change itself seems reasonable though.

**Comment by simonpuchert** [comment]
I haven't looked deep into the algorithm, but an alternative would be using the norm of the candidate (before it's normalized, of course) to estimate the distance from the normalized arithmetic mean (the start) to the spherical mean, i.e. something like inserting at the end of step 1: `if (norm > 1-epsilon) return center + rho*candidate;`.

The idea is, if the arithmetic mean is near the sphere, the weight has to be concentrated.

On the topic of proving bounds, by analyzing the projection of the arithmetic mean on the tangential space at the spherical mean, I can show that for max_distance < epsilon, the maximum error is roughly epsilon^3 / 24. I don't know how to typeset in Github markdown, but the proof is rather easy anyway (Hint: sup(|sin(x) - (1 - epsilon^2 / 8) * x|, x in [0, epsilon]) = epsilon^3 / 24 + O(epsilon^4)).

**Comment by kronbichler** [comment]
Thanks @simonpuchert for investigating this issue. Given this estimate, we know that the error is definitely not larger than `2.5e-8` with the threshold at 0.01. I think this is safe enough also regarding my third item above, given that we never really hit this worst case apparently. @masterleinad can you add this comment to the implementation of the short path so we remember this?

After that, let's wait for the investigation by @gassmoeller and then merge.

**Comment by masterleinad** [comment]
@simonpuchert Can you elaborate a bit?

The arithmetic mean `x_a` satisfies `x_a = w_i\sum_i d_i` while for the spherical mean `x_s` it holds `x_s = w_i \sum_i p_{x_s}(d_i)` where `p_{x_s}` is the projection to the tangential plane. This implies `|x_a - x_s|\leq \max_i |d_i-p_{x_s}(d_i)|`. Now it seems safe to assume that `max_distance<epsilon` also implies `|x_s - d_i|<epsilon`. For sufficienty small `epsilon`, the maximum distance between `x_s` and `p_{x_s}(d_i)` can be approximated by `epsilon+epsilon^3/24+O(\epsilon^5)`. This implies that the distance between `p_{x_s}(d_i)` and `d_i` can be bounded by `epsilon^2/2+epsilon^4/36`. Choosing `epsilon=sqrt(1.e-2)` as we currently do, I end up with an estimate of magnitude `5.e-3` which is far away from the tolerance `1.e-10` we want to gurantee.

**Comment by simonpuchert** [comment]
You have to analyze the (orthogonally) projected arithmetic mean, not the mean itself. Admittedly, I didn't explain it well (or even at all).

Let p be the _orthogonal projection_ on the tangential space at the spherical mean (defined so that x_s gets mapped to 0, in essence it's the projection onto Clocalx, Clocaly coordinates just like in the implementation).

Because the spherical mean is our reference point, we have
0 = p(x_s) = sum( w_i * theta[i] * p(d_i) / |p(d_i)| ).

On the other hand,
p(x_a) = sum( w_i * sin(theta[i]) * p(d_i) / |p(d_i)| ).

Subtracting the projection of the spherical mean ( = 0 ), we obtain
p(x_a) = sum( w_i * (sin(theta[i]) - (1-epsilon^2/8)*theta[i]) * p(d_i) / |p(d_i)| )

The triangle inequality yields
|p(x_a)| <= bound on (sin(theta[i]) - (1-epsilon^2/8)*theta[i]) ~ epsilon^3 / 24.

The bound on the normalized arithmetic mean (the starting value) itself is pretty much the same and the projection doesn't substantially reduce the distance since our whole scenario is contained within a small sector of the sphere.

I hope the "proof" is somewhat understandable now. The bound is asymptotically optimal, but the worst case consists of collinear points.

**Comment by masterleinad** [comment]
OK. I still don't quite understand your estimates.

x_s can be written as x_s=sum_i w_i p_s(d_i) where p_s is the inverse exponential map.
Now, the orthogonal projection of a point x to the plane tangential to x_s can be wriiten as
p(x) = x-((x-x_s)*x_s)*x_s/|x_s|^2 (of course |x_s|=1 in the current setting)
This implies
x_s-p(x_a)=sum_i w_i((d_i-x_s)-((d_i-x_s)*x_s)*x_s)*theta_i
where theta_i is the angle between d_i, the center and x_s. Assuming |x_s-d_i|<epsilon theta_i can be estimated by epsilon.
(d_i-x_s)*x_s=|d_i-x_s|*sin(theta_i/2)<=|d_i-x_s|*(epsilon/2+O(epsilon^3))
and from the estimates in my previous comment
|d_i-x_s|<epsilon^2/2 +O(epsilon^4).
In summary, this means
|x_s-p(x_a)|<epsilon^3/2+O(epsilon)^4.

Can you give some more details why 
p(x_a) = sum( w_i * sin(theta[i]) * p(d_i) / |p(d_i)| )
holds and why this implies (how comes 1-epsilon^2/8 into play?)
p(x_a) = sum( w_i * (sin(theta[i]) - (1-epsilon^2/8)*theta[i]) * p(d_i) / |p(d_i)| )
?

**Comment by masterleinad** [comment]
Closed in favor of #5562.

**Comment by simonpuchert** [comment]
The equation 
p(x_a) = sum( w_i * sin(theta[i]) * p(d_i) / |p(d_i)| )
stems from the definition x_a = sum( w_i * d_i ), the equation |p(d_i)| = sin(theta[i]) and the linearity of the orthogonal projection p(x) = x - (x*x_s)*x_s.
Your map, while one would be more inclined to call it "the orthogonal projection", is not linear (since p(0) != 0). Another thing I didn't explain ... at least it should be clear now.

The other line is the subtraction p(x_a) - (1-epsilon^2/8) p(x_s) = p(x_a)
and the factor (1-epsilon^2/8) is chosen to (almost) minimize the supremum of the terms in the parentheses.
inf{sup{|sin(x) - c*x|, x in [0, epsilon]}, c in R} is assumed for c ~ 1 - epsilon^2 / 8 and is ~ epsilon^3 / 24.
